### PR TITLE
Add initial brigadier support (implements #77)

### PR DESCRIPTION
**Pull request by MiniDigger**
~this is by no means ready~, just created a place for open points (will not all be tackled in this PR)

- [x] register commands to brigadier
- [x] register params + primitive types to brigadier
- [x] provide static tab completions
- [x] provide dynamic tab completions
- [ ] more complex getArgumentTypeByClazz (support min, max etc)
- [ ] custom argument types (huge stretch) 
- [x] clean up maven dependencies again
- [x] clean up api
- [x] check if it works with perms
- [x] check command blocks

implements #77

this is affected by https://bugs.mojang.com/browse/MC-165562

### how to use:
you just call `commandManager.enableUnstableAPI("brigadier");` on a paper command manager

### CODE REVIEW

**Comment by MiniDigger** [comment]
18:36 <+Choco> But the methods you want to register an argument type is ArgumentRegistry.a(new MinecraftKey("namespace:for_argument_type"), YourArgumentType.class, new ArgumentSerializerVoid(SupplerForYourArgumentType));
18:36 <+Choco> As an example, ArgumentRegistry.a(new MinecraftKey("veinminer:tool_type"), ToolArgumentType.class, new ArgumentSerializerVoid<>(ToolArgumentType::tool));

File: bukkit/src/main/java/co/aikar/commands/BukkitBrigadierManager.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+public class BukkitBrigadierManager<S> extends ACFBrigadierManager<S> {
+
+    public BukkitBrigadierManager(CommandManager<?, ?, ?, ?, ?, ?> manager, CommandDispatcher<S> dispatcher) {
+        super(manager, dispatcher);
+
+        //TODO custom argument types?
+//        registerArgument(Player.class, new ArgumentType<Player>() {
```
Comment (aikar) [inline bukkit/src/main/java/co/aikar/commands/BukkitBrigadierManager.java]: Def not going to do a manual brigadier mapping to every context. we have to map automatically on every registered context.

**Comment by MrIvanPlays** [comment]
Any activity on that ?

**Comment by aikar** [comment]
Some general notes/desires.

What I was hoping for is for this module to have no bukkit references. purely Brigadier Root/Literal/Argument CommandNode references.

Then, I want ACF to Add respective CommandNode children to the RootCommandNode as defined by ACF's metadata.

Then we would add an event in Paper PlayerBuildBrigadierCommands that exposes the RootCommandNode, and ACF then manipulates it (adding children)

Then, the same event can be added to Sponge or any platform using brigadier and this entire module can be utilized over all of them as a shared core.

As I commented on one line, we shouldnt have any kind of direct object type mapping. Just expose the Argument name and basic primitive types (numbers/strings) as its ArgumentNodes

File: example/pom.xml [hunk 2]
```diff
@@ -64,10 +64,10 @@
...
                     <dependencyReducedPomLocation>${project.build.directory}/dependency-reduced-pom.xml</dependencyReducedPomLocation>
                     <relocations>
-                        <relocation>
-                            <pattern>co.aikar.commands</pattern>
-                            <shadedPattern>com.exampleplugin.acf</shadedPattern>
-                        </relocation>
+                        <!--<relocation>-->
+                            <!--<pattern>co.aikar.commands</pattern>-->
+                            <!--<shadedPattern>com.exampleplugin.acf</shadedPattern>-->
+                        <!--</relocation>-->
```
Comment (MiniDigger) [inline example/pom.xml:70]: gotta remove this before merging

**Comment by MiniDigger** [comment]
https://i.imgur.com/oyaKVVL.gifv

**Comment by MiniDigger** [comment]
this depends now on https://github.com/PaperMC/Paper/pull/3193 and is thus Paper only

**Comment by MiniDigger** [comment]
now supports perms https://i.imgur.com/Jbi58QL.gifv

**Comment by MiniDigger** [comment]
I would say this is ready for review no 🎉

**Comment by Sauilitired** [comment]
Imgur
Imgur

```java
@Subcommand("find|where") @CommandPermission("hyperverse.find") @CommandAlias("hvf|hvfind") @CommandCompletion("@players") @Description("{@@command.find}") public void findPlayer(final CommandSender sender, final String player) { /* stub */ }
```
It fails on the command alias, but works fine on the subcommand. I am just completing my own name, so the tab completion works fine in the second example. Both commands execute the same, so it's purely a Brigadier thing.

**Comment by MiniDigger** [comment]
@Sauilitired hope you happy now https://i.imgur.com/hKZgVQg.gifv

**Comment by MiniDigger** [comment]
added support for greedy strings https://i.imgur.com/MV64OyJ.gifv

**Comment by MiniDigger** [comment]
https://i.imgur.com/VtUaYrm.png

**Comment by MiniDigger** [comment]
@aikar brigadier now enables automatically if you call `commandManager.enableUnstableAPI("brigadier");` on a **Paper** command manager. 

I currently dont see how we can make the bukkit command manager enable paper stuff when it detects its being used on paper

**Comment by darbyjack** [comment]
> @aikar brigadier now enables automatically if you call `commandManager.enableUnstableAPI("brigadier");` on a **Paper** command manager.
> 
> I currently dont see how we can make the bukkit command manager enable paper stuff when it detects its being used on paper

So if I understand this correctly, if I'm using Paper Manager (which will fallback to Bukkit if they aren't on Paper) and they ARE on Paper, and on a support version for Brigadier, it will attempt to utilize it? Else it will fallback to Paper?

**Comment by MiniDigger** [comment]
nvm am dum, thanks glare. it does that now

I think that was the last thing I wanted to work on! Soooo

:shipit:

File: paper/pom.xml [hunk 3]
```diff
@@ -64,5 +64,11 @@
             <version>1.15.2-R0.1-SNAPSHOT</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>com.destroystokyo.paper</groupId>
+            <artifactId>paper-mojangapi</artifactId>
+            <version>1.15.2-R0.1-SNAPSHOT</version>
+            <scope>provided</scope>
+        </dependency>
```
Comment (MiniDigger) [inline paper/pom.xml:72]: duplicated dependency @aikar

**Comment by NiklasEi** [comment]
Nice, thanks for the work!
Is it planned at some point to not only support this for Paper?

**Comment by JOO200** [comment]
It should be possible to expose brigadier stuff to waterfall too.
For other server implementations it depends on the server architecture or if they expose brigadier like paper

**Comment by MrIvanPlays** [comment]
> It should be possible to expose brigadier stuff to waterfall too.
> For other server implementations it depends on the server architecture or if they expose brigadier like paper

There's a discussion about it in bungeecord, here: https://github.com/SpigotMC/BungeeCord/issues/2812

**Comment by MiniDigger** [comment]
if you mean support for spigot: that is up to whatever spigot decides to do. I know that I will not hack my way around the limitations spigot introduces (and I doubt aikar will), but if spigot decides to adapt papers api (or something similar), we could consider switching.

but like, this is yet another reason for server admins to upgrade to paper.

**Comment by MrIvanPlays** [comment]
> custom argument types (huge stretch)

_cough_ mojang set limitation _cough_

**Comment by aikar** [comment]
since this is marked unstable, im going to merge for now, as the changes to fix the issues i saw really dont risk plugins and plugins can control if they opt into this or not.

note that issuer aware context resolvers may have funky results.

**Pull request merged**